test/unit/CowOperator/closeOrder.js

```diff
@@ -0,0 +1,368 @@
+
+      expect(initialPoolEth).to.be.gt(0);
+      expect(initialOperatorWeth).to.be.gt(0);
+
+      await swapOperator.closeOrder(contractOrder);
+
+      expect(await ethers.provider.getBalance(pool.address)).to.eq(initialPoolEth.add(initialOperatorWeth));
+      expect(await weth.balanceOf(swapOperator.address)).to.eq(0);
+    });
+
+    it('transfers the erc20 token to pool if its not weth', async function () {
+      // Cancel current order
+      await swapOperator.closeOrder(contractOrder);
+
+      // Advance time to enable swapping again
+      await time.increase(MIN_TIME_BETWEEN_ORDERS);
+
+      // Place an order swapping DAI for ETH
+      const { newOrder, newContractOrder, newOrderUID } = await setupSellDaiForEth({ validTo: await lastBlockTimestamp() + 650 });
+      await weth.mint(cowVaultRelayer.address, order.buyAmount);
+      await swapOperator.placeOrder(newContractOrder, newOrderUID);
+
+      const checkpointPoolDai = await dai.balanceOf(pool.address);
+      const checkpointOperatorDai = await dai.balanceOf(swapOperator.address);
+      expect(checkpointOperatorDai).to.eq(newOrder.sellAmount.add(newOrder.feeAmount));
+
+      // Add some extra sellToken to the swap operator balance, simulating a token rebase
+      const extraAmount = parseEther('1');
+      await dai.mint(swapOperator.address, extraAmount);
+
+      await swapOperator.closeOrder(newContractOrder);
+
+      expect(await dai.balanceOf(pool.address)).to.eq(checkpointPoolDai.add(checkpointOperatorDai).add(extraAmount));
+      expect(await dai.balanceOf(swapOperator.address)).to.eq(0);
+    });
+  });
+
+  describe('emitting OrderClosed event', function () {
+    it('when order was not filled', async function () {
+      const tx = await swapOperator.closeOrder(contractOrder);
+      const rcp = await tx.wait();
+
+      const event = rcp.events.find(e => e.event === 'OrderClosed');
+
+      expect(event.args.order).to.deep.include.members(Object.values(contractOrder));
+      expect(event.args.filledAmount).to.eq(0);
+    });
+
+    it('when order was partially filled', async function () {
+      await cowSettlement.fill(contractOrder, orderUID, order.sellAmount.div(2), order.feeAmount.div(2), order.buyAmount.div(2));
+
+      const tx = await swapOperator.closeOrder(contractOrder);
+      const rcp = await tx.wait();
+
+      const event = rcp.events.find(e => e.event === 'OrderClosed');
+
+      expect(event.args.order).to.deep.include.members(Object.values(contractOrder));
+      expect(event.args.filledAmount).to.eq(order.sellAmount.div(2));
+    });
+
+    it('when order was fully filled', async function () {
+      await cowSettlement.fill(contractOrder, orderUID, order.sellAmount, order.feeAmount, order.buyAmount);
+
+      const tx = await swapOperator.closeOrder(contractOrder);
+      const rcp = await tx.wait();
+
+      const event = rcp.events.find(e => e.event === 'OrderClosed');
```

Comment:
Hi! I'm curious whether you were aware of this or if you've encountered any inconveniences using it:

https://ethereum-waffle.readthedocs.io/en/latest/matchers.html?highlight=events#emitting-events

Reply:
Hi! Yes, I'm aware of that matcher but was not able to properly match the arguments of the event. So the closest I got using that was: `await expect(swapOperator.closeOrder(contractOrder)).to.emit(swapOperator, 'OrderClosed').withArgs(Object.values(contractOrder), 0);` , but it doesn't work because the ordering of the order attributes is different. If I had a more powerful argument matcher I'd do `withArgs(arrayMatching(Object.values(contractOrder)), 0)`, but afaik we don't have that available (please correct me if I'm wrong). So instead of manually constructing the expectation (tuple with 12 elements), I first got the receipt and used an available matcher: `expect(event.args.order).to.deep.include.members(Object.values(contractOrder));`

Reply:
I see. Curently node seems to sort the values in the order they are declared in the object so simply moving the receiver key-value pair right after buyToken in order would make the test pass using that matcher, but as far as I know the order of Array.values which is the same as for...in should not be relied upon as it's implementation specific. (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in)

My initial nitpicky concern towards using chai's include was that two keys could theoretically end up having the same value. It is however highly unlikely since the order comes from calldata and nothing alters it. 

But in other words, if someone were to trick the test on purpose it would be possible by modifying one of the properties in the Data struct from the emitted event such as feeAmount, making it equal to sellAmount for example, and the test would pass even though the order object is not the same as the one provided as a parameter.

If you think it's worth it you just need to get the values in the correct order by using a function that maps the ordered Data struct keys to the values in the object.

Reply:
I see your point. Well, to be 100% safe I just updated the test to reflect your suggestion, now I build the order tuple with a function and use it for an exact match using `withArgs` matcher.